lib/core/src/display/navigation/event_handler.rs

```diff
@@ -0,0 +1,326 @@
+        let ehandler = Rc::new(Self { data });
+
+        let document : EventTarget = dyn_into(document().unwrap()).unwrap();
+        let target   : EventTarget = dyn_into(dom.dom.clone()).unwrap();
+
+        ehandler.initialize_events(&target, &document)?;
+        Ok(ehandler)
+    }
+
+    /// Initialize mouse events.
+    fn initialize_events
+    (self:&Rc<Self>, target:&EventTarget, document:&EventTarget) -> Result<()> {
+        ignore_context_menu(&target)?.forget();
+        self.initialize_mouse_start(&target)?;
+        self.initialize_mouse_move(&target)?;
+        self.initialize_mouse_end(&target, &document)?;
+        self.initialize_wheel_event(&target)?;
+        Ok(())
+    }
+
+    /// Initialize MouseDown event callback.
+    fn initialize_mouse_start
+    (self:&Rc<Self>, target:&EventTarget) -> Result<()> {
+        let ehandler = self.clone();
+        let closure = move |event:MouseEvent| {
+            if event.button() == MMB || event.button() == RMB {
+                let mut ehandler = ehandler.data.borrow_mut();
+                let button       = event.button();
+                let x            = event.x() as f32;
+                let y            = event.y() as f32;
+                let position     = Vector2::new(x, y);
+                let start        = MouseStart { button, position };
+                *ehandler        = InternalMouseEvent::Start(start);
+            }
+        };
+        add_mouse_event(&target, "mousedown", closure)?.forget();
+        Ok(())
+    }
+
+    /// Initialize MouseMove event callback.
+    fn initialize_mouse_move
+    (self:&Rc<Self>, target:&EventTarget) -> Result<()> {
+        let ehandler = self.clone();
+        let closure = move |event:MouseEvent| {
+            let ime     = &mut ehandler.data.borrow_mut();
+            let ime     = ime as &mut InternalMouseEvent;
+            let x       = event.x() as f32;
+            let y       = event.y() as f32;
+            let current = Vector2::new(x, y);
+            match ime {
+                InternalMouseEvent::Start(mouse_start) => {
+                    let previous = mouse_start.position;
+
+                    let start      = Some(mouse_start.clone());
+                    let mouse_move = MouseMove { start, previous, current };
+                    *ime = InternalMouseEvent::Move(mouse_move);
+                },
+                InternalMouseEvent::Move(mouse_move) => {
+                    mouse_move.current = current;
+                },
+                InternalMouseEvent::Wheel(mouse_wheel) => {
+                    mouse_wheel.position = current;
+                },
+                _ => {
+                    let start      = None;
+                    let previous   = current;
+                    let mouse_move = MouseMove { start, previous, current };
+                    *ime = InternalMouseEvent::Move(mouse_move);
+                },
+            }
```

Comment:
mut borrowing from RefCell for such a large block is rather a bad design decision. Refcell's mut borrows should be short and very scoped.

Reply:
How can it be done differently if I am matching it?

Reply:
I don't know. It just looks bad. Maybe it is needed this way but I doubt it. To properly answer your question I would need to do proper review again having OK docs.